Suppress stack trace in test

som-snytt opened the pull request.

som-snytt committed:
Suppress stack trace in test

hrhino commented on src/testkit/scala/tools/testkit/AssertUtil.scala:
```diff
@@ -201,9 +201,41 @@ object AssertUtil {
   def readyOrNot(awaitable: Awaitable[_]): Boolean = Try(Await.ready(awaitable, TestDuration.Standard)).isSuccess
+
+  /** Syntax to clarify which is the expected value. */
+  def assertEqualTo[A](expected: A)(actual: A): Unit = assertEquals(expected, actual)
```
This cries for a shiny new extension method: `(1 + 1) assertEqualTo 2`...

som-snytt replied:
Don't make me dig out that meme with the red dress.

We recently re-watched The Matrix, which I don't adore as some do, but the daughter said we watched it already, and then oh that's where that meme comes from, meaning the red/blue pill. I can never remember which pill is which. Anyway, the red dress from Matrix is a thing but I think not a meme? This red dress is the distraction that leads to destruction. It was conceived as a test, but the coder guy says he has recycled it for other temptations.

hrhino commented on src/testkit/scala/tools/testkit/AssertUtil.scala:
```diff
@@ -201,9 +201,41 @@ object AssertUtil {
+    thread.join()
+  }
+
+  private def addError(t: Thread, e: Throwable): Unit = synchronized(errors.+=((t, e)))
```
what do you have against infix?

som-snytt replied:
Honestly, I always forget whether infix will work or it takes tuples or multiargs wrong.

dwijnand replied:
Double-up parens to be sure (to be sure)

```scala
errors += ((t, e))
```

hrhino commented on test/junit/scala/sys/process/ProcessTest.scala:
```diff
@@ -102,14 +103,23 @@ class ProcessTest {
+      }
+    }
+    stackless.run()
+    stackless.assert()
```
I'd have made this part of `.run()`, but this is more separation-of-concerns-y.

dwijnand replied:
Yeah if the intended usage is to run and assert, I'd make `run` assert, and have a more verbose `runWithoutAssert` path-least-taken option.  (But not a blocker..)

hrhino reviewed (approved):
Testing PRs never get any love.

dwijnand commented on src/testkit/scala/tools/testkit/AssertUtil.scala:
```diff
@@ -201,9 +201,41 @@ object AssertUtil {
+
+  def getErrors: List[(Thread, Throwable)] = synchronized(errors.toList)
+
+  def assert(): Unit = getErrors.collect { case (_, e: AssertionError) => e }.headOption.foreach(e => throw e)
```
Maybe it's best to report them all?

```suggestion
  def assert(): Unit = {
    getErrors.collect { case (_, e: AssertionError) => e } match {
      case Nil      => ()
      case e :: Nil => throw e
      case errors   =>
        val acc = new AssertionError("Multiple assertion errors")
        errors.foldLeft(acc) { (acc, err) =>
          acc.addSuppressed(err)
          acc
        }
        throw acc
    }
  }
```

Or, if you prefer "last exception wins" semantics (like the suppression in `try throw X finally throw Y`):

```suggestion
  def assert(): Unit = {
    getErrors.collect { case (_, e: AssertionError) => e } match {
      case Nil    => ()
      case errors => throw errors.reduceRight { (err, acc) => acc.addSuppressed(err); acc }
    }
  }
```

dwijnand commented on src/testkit/scala/tools/testkit/AssertUtil.scala:
```diff
@@ -201,9 +201,41 @@ object AssertUtil {
   def readyOrNot(awaitable: Awaitable[_]): Boolean = Try(Await.ready(awaitable, TestDuration.Standard)).isSuccess
+
+  /** Syntax to clarify which is the expected value. */
+  def assertEqualTo[A](expected: A)(actual: A): Unit = assertEquals(expected, actual)
```
Is it also part of the intent to make them not LUB?  And/or is it so the second one can be a block?

dwijnand reviewed (approved):
LGTM too.  A couple of ideas/suggestions.

som-snytt committed:
Use background thread handling in tests

Leverage the code that waits for background threads
to die, when managing any assertion errors they throw.

Test code gets to examine the result of the main block
if it completed, as well as the collected exceptions
thrown by all threads in the test group. If the test
code does not signal failure by returning an exception
to throw, the rig checks for assertion errors.

som-snytt commented:
I returned some ❤️ and incorporated all the suggestions. I also gave it a better pun, `withoutATrace`.

I'm not sure if it will be hugely useful, but I think I wind up needing something like it on every project. 

Note you don't see suppressed in JUnit 4, but maybe in JUnit 5.

`sys.process` is uniquely perverse in kicking off threads. Usually I'd expect to run tasks on a pool, so that might require a bit more work, to be useful.

I have doubts about `assertEqualsTo` because these methods are always heavily overloaded for specialization, and that's probably not worth it here. I'll probably revert that, or let the maintainer of `sys.process` module do it.

som-snytt committed:
Revert speculative assert syntax

lrytz commented:
Thank you all!

som-snytt commented:
The force push included a refactoring, but you know, it's only test code. It will get another review when the build starts failing intermittently.
